chain/chain/src/chain.rs

```diff
@@ -2106,6 +2106,52 @@ impl Chain {
         Ok(new_head)
     }
 
+    /// Update flat storage for given processed or caught up block.
+    fn update_flat_storage_for_block(
+        &mut self,
+        block: &Block,
+        shard_id: ShardId,
+    ) -> Result<(), Error> {
+        if let Some(flat_storage_state) =
+            self.runtime_adapter.get_flat_storage_state_for_shard(shard_id)
+        {
+            let mut new_flat_head = *block.header().last_final_block();
+            if new_flat_head == CryptoHash::default() {
+                new_flat_head = *self.genesis.hash();
+            }
+            // Try to update flat head.
+            flat_storage_state.update_flat_head(&new_flat_head).unwrap_or_else(|err| {
+                match &err {
+                    FlatStorageError::BlockNotSupported(_) => {
+                        // It's possible that new head is not a child of current flat head, e.g. when we have a
+                        // fork:
+                        //
+                        //      (flat head)        /-------> 6
+                        // 1 ->      2     -> 3 -> 4
+                        //                         \---> 5
+                        //
+                        // where during postprocessing (5) we call `update_flat_head(3)` and then for (6) we can
+                        // call `update_flat_head(2)`. In such case, just log an error.
+                        debug!(target: "chain", "Cannot update flat head to {:?}: {:?}", new_flat_head, err);
+                    }
+                    _ => {
+                        // All other errors are unexpected, so we panic.
+                        panic!("Cannot update flat head to {:?}: {:?}", new_flat_head, err);
+                    }
+                }
+            });
+        } else {
+            // If flat storage doesn't exist, update its creation status.
+            match &mut self.flat_storage_creator {
+                Some(flat_storage_creator) => {
+                    flat_storage_creator.update_status(shard_id, &self.store)?;
```

Comment:
Maybe it will be less confusing and also more robust to move the call to `flat_storage_creator.update_status` to outside of block processing. Technically, we don't need to call this function every time when we process a block. We only need to call it periodically to update the process of flat state migration. And if the node is not processing any new blocks (if it is in state sync), that means flat storage creation will not make progress either. I'd suggest creating a new event in ClientActor that is triggered periodically (say every 100ms) and calls `update_status` (see https://github.com/near/nearcore/blob/63d77d49abf6ad5cda3ae6ad1488c33efda3b342/chain/client/src/client_actor.rs#L1112).

Reply:
ah are you worried about the potential race conditions between the triggers of `update_status` and `save_flat_state_changes`. I think the scenario you described won't happen because we are not moving `update_status` to a different thread, it is still inside `ClientActor`. It's just triggered differently. Instead of triggered at block processing time, it is triggered periodically every 100ms.

Reply:
Oh, for some reason I thought that we can save flat state changes in a thread spawned for applying chunk, but it always happens in main thread. And I didn't realize that catchup is also there, only the heavy work is moved outside.
So I'll just try to move `update_status` as you suggested.

Reply:
Created an issue for that: https://github.com/near/nearcore/issues/8250